### ISSUE TITLE
Incorrect quantile computation

### ISSUE THREAD

**Issue by jverre-drivalytix** [simple-statistics/simple-statistics#353]
While looking at the quantile computation, I have noticed that the quantile calculation does not match up with the calculations returned by numpy in python (can be considered as the reference).

There are 4 different ways to interpolate quantiles when one quantile does not land on an exact value:
* linear
* lower
* higher
* midpoint
* nearest

Using data = [0, 0, 0.3, 1.2, 1.23,  3.5, 10, 12, 23.3, 32.1] and computing the 25th, median and 75th percentile we get:
* simple-statistics: [ 0.3, 2.365, 12.0 ]
* numpy / linear: [ 0.525,  2.365, 11.5  ]
* numpy / lower: [ 0.3,   1.23, 10.0  ]
* numpy / higher: [ 1.2,  3.5, 12.0 ]
* numpy / midpoint: [ 0.75,   2.365, 11.0   ]
* numpy / nearest: [ 0.3,   1.23, 12.0  ]

For reproducibility, this is the code I used:
```
const ss = require('simple-statistics')

var data = [0, 0, 0.3, 1.2, 1.23,  3.5, 10, 12, 23.3, 32.1]
console.log(ss.quantile(data, [0.25, 0.5, 0.75]))
```

```
import numpy as np
data = [0, 0, 0.3, 1.2, 1.23,  3.5, 10, 12, 23.3, 32.1]
for i in ['linear', 'lower', 'higher', 'midpoint', 'nearest']:
    print(i, np.percentile(data, [25, 50, 75], interpolation=i))
```

**Comment by tmcw** [comment]
We're currently using quickselect for quantiles, so unfortunately it's a little complicated. I'll take a look at the implementation, see if there's a fix, and potentially replace it with a simpler (albeit less performant) implementation.

**Comment by mourner** [comment]
@tmcw I think quickselect doesn't matter here. What defines the behavior is [this logic](https://github.com/simple-statistics/simple-statistics/blob/master/src/quantile_sorted.js#L14-L38) for sorted quantile. We seem to use some kind of a mix between Python's equivalent `nearest` and `midpoint` depending on whether the index landed on an integer value.

**Comment by Yomguithereal** [comment]
I agree that we should stick to numpy's behavior but just know that there still are controversies about their choices which are currently being discuted by Python's core concerning the addition of the `quantile` methods to the py3 `statistics` module right now.

**Comment by mhkeller** [comment]
Any update on this?

**Comment by tmcw** [comment]
I think we should switch from the blend of nearest and midpoint to just one or the other. Reading through the python [documentation for quantiles](https://docs.python.org/3/library/statistics.html#statistics.quantiles) it seems like they shipped a version with _just_ `nearest`-like behavior. Would welcome any other ecosystem examples for what we should do here.

It's tempting to support a linear/nearest/etc option so that this change doesn't have to be a major version bump, but if the current behavior isn't what anyone would want as a default, the change to fix this issue will be a major version bump.

**Comment by yiyange** [comment]
somewhat related since it is part of the same function, I am not sure about the last two conditions
https://github.com/simple-statistics/simple-statistics/blob/1db09fc5d781fb661a2d0c7b413e3acead531696/src/quantile_sorted.js#L29-L36

in what cases will we actually get there?

if `p` is smaller than 0 or larger than 1, it errs; 

if `p` is 0 or 1 (the only possible integers left), it returns; 

if `idx` is non-int, the same as saying `p` is non-int since `x.length` is integer (which is the only possibility left, which makes `idx % 1 !== 0` a bit futile), it returns.

Am I missing any remaining possibility?

**Comment by yiyange** [comment]
Somewhat related, this line of comment is a bit misleading (i was mislead by this thus made some incorrect comment)

https://github.com/simple-statistics/simple-statistics/blob/1db09fc5d781fb661a2d0c7b413e3acead531696/src/quantile_sorted.js#L27

p is definitely not an integer by this point. but `idx` can be integer or a float

**Comment by rbox-risk** [comment]
@tmcw From the long and well established R stats ecosystem, the [quantile](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html) function runs 9 different algorithms, of which `type=7` is the default. This default agrees with the numpy / linear output with these data.

```
rdata <- c(0, 0, 0.3, 1.2, 1.23, 3.5, 10, 12, 23.3, 32.1)
quantile(rdata, c(0.25,0.5,0.75))
```
gives

```
   25%    50%    75% 
 0.525  2.365 11.500 
```
Running all types
```
t(sapply(1:9, function(i) quantile(rdata, c(0.25,0.5,0.75), type=i)))
```
gives output (with annotations):
```
          25%   50%      75%
 [1,] 0.30000 1.230 12.00000 (ss python)
 [2,] 0.30000 2.365 12.00000
 [3,] 0.00000 1.230 12.00000
 [4,] 0.15000 1.230 11.00000
 [5,] 0.30000 2.365 12.00000 (ss)
 [6,] 0.22500 2.365 14.82500
 [7,] 0.52500 2.365 11.50000 (R default, numpy default/linear)
 [8,] 0.27500 2.365 12.94167
 [9,] 0.28125 2.365 12.70625
```
recovering the ss python implementation as `type=1` and the ss js implementation as `type=5`. The documentation describes each algo, and provides some useful [references ](https://blogs.sas.com/content/iml/2017/05/24/definitions-sample-quantiles.html)

**Comment by isaiah2004** [comment]
> I think we should switch from the blend of nearest and midpoint to just one or the other. Reading through the python [documentation for quantiles](https://docs.python.org/3/library/statistics.html#statistics.quantiles) it seems like they shipped a version with _just_ `nearest`-like behavior. Would welcome any other ecosystem examples for what we should do here.
> 
> It's tempting to support a linear/nearest/etc option so that this change doesn't have to be a major version bump, but if the current behavior isn't what anyone would want as a default, the change to fix this issue will be a major version bump.

do we need the major version bump?

**Comment by MohammedAlkindi** [comment]
This got fixed, but the thread never closed. #783 (merged 2026-03-10, shipped in **v7.8.9**) switched `quantileSorted` to type-7 linear interpolation — the numpy/R default this thread converged on. The repro from the issue body, on today's `main` (8449cf5):

```js
const data = [0, 0, 0.3, 1.2, 1.23, 3.5, 10, 12, 23.3, 32.1];
ss.quantile(data, [0.25, 0.5, 0.75]);               // [0.5249999999999999, 2.365, 11.5]
[0.25, 0.5, 0.75].map((p) => ss.quantile(data, p)); // same three values
```

That is the `numpy / linear: [0.525, 2.365, 11.5]` row from the issue body, up to floating point, and R's default `type=7` in the annotated comparison above. The array and scalar call forms now agree with each other too (#805, shipped in 7.9.2, with a regression test keyed to #704), and the misleading `p is integer` comment flagged above was rewritten as part of #783.

Looks closable.
